[58105] Added update_item and delete_branch methods to HierarchicalItemService

apfohl opened the pull request:
# Ticket
https://community.openproject.org/projects/openproject/work_packages/58105/github

# What are you trying to accomplish?
See headline

# What approach did you choose and why?
same as last time

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

Commit:
[58105] Added update_item and delete_branch methods to HierarchicalItemService

Kharonus commented on app/contracts/custom_fields/hierarchy/update_item_contract.rb:
```diff
@@ -0,0 +1,61 @@
+      end
+
+      rule(:item) do
+        if value.is_a?(CustomField::Hierarchy::Item)
```
🔴 @mereghost Did you update this here, too? In regard of checking for the type in the schema already?

Kharonus commented on app/contracts/custom_fields/hierarchy/update_item_contract.rb:
```diff
@@ -0,0 +1,61 @@
+        next unless key?
+
+        if CustomField::Hierarchy::Item.exists?(parent_id: values[:item].parent, label: value)
+          key.failure("Label must be unique within the same hierarchy level")
```
🟢 I will put this message on localization once I start with the update action in the UI. The localization will get merged today.

Kharonus commented on app/models/custom_field/hierarchy/item.rb:
```diff
@@ -32,5 +32,5 @@ class CustomField::Hierarchy::Item < ApplicationRecord
 
   belongs_to :custom_field
-  has_closure_tree order: "sort_order", numeric_order: true
+  has_closure_tree order: "sort_order", numeric_order: true, dependent: :delete_all
```
👍🏽

Kharonus commented on app/services/custom_fields/hierarchy/hierarchical_item_service.rb:
```diff
@@ -55,7 +55,22 @@ def insert_item(parent:, label:, short: nil)
+      def delete_branch(item:)
+        return Failure(:item_is_root) if item.root?
+
+        # CustomField::Hierarchy::Item sets "dependent: :destroy"
```
🟡 can we remove the comment? It is already oudated :D

mereghost replied:
No! :P Sure.

Kharonus commented on app/services/custom_fields/hierarchy/hierarchical_item_service.rb:
```diff
@@ -67,13 +82,18 @@ def create_root_item
-        item = CustomField::Hierarchy::Item
-                 .create(parent: validation[:parent], label: validation[:label], short: validation[:short])
+      def create_child_item(validation:)
+        item = validation[:parent].children.create(label: validation[:label], short: validation[:short])
```
🟡 Hmm, not sure, but I do not like this change. What is the benefit? Before the code was very clear, that a new Item is created. Now I must understand, that the item to create is referenced here as `.children`.

🔴 Does this even work, if the created item is not the only child?

mereghost replied:
Because that's how `closure_tree` works. It keeps intermediate representations of the tree for lookup, so we need to go through its code.

From another perspective, it is clearer to the random RoR dev as `children.create` maps to "I'm creating some associated record to the caller". :P

Kharonus replied:
Can you confirm, that this does not break, if the parent already has persisted children, when adding a new one?

mereghost replied:
Yes. It works. Why? Because `.children` is an association proxy that don't override anything but ensure that the fields for the association are correctly set - and in case of closure tree, sets up the rest of the fields and optimizations.

Kharonus commented on spec/services/custom_fields/hierarchy/hierarchical_item_service_spec.rb:
```diff
@@ -63,39 +65,99 @@
-        allow(CustomField::Hierarchy::Item)
-          .to receive(:create).and_return(instance_double(CustomField::Hierarchy::Item,
-                                                          persisted?: false, errors: "some errors"))
+        # rubocop:disable RSpec/VerifiedDoubles
```
❓ This cop should not trigger, as `instance_double` is a verified double?

mereghost replied:
But the wrapping double isn't. I'll see if I can remove this by adding the correct double here.

mereghost replied:
Fixed.

mereghost committed:
Clean up some contracts adds get_branch

# Conflicts:
#	app/contracts/custom_fields/hierarchy/insert_item_contract.rb
#	spec/contracts/custom_fields/hierarchy/insert_item_contract_spec.rb

mereghost committed:
Update specs, contracts, draft inplementation of the missing actions

Kharonus commented on app/contracts/custom_fields/hierarchy/generate_root_contract.rb:
```diff
@@ -32,11 +32,7 @@ module CustomFields
-
-      rule(:hierarchy_root) do
-        key.failure("Hierarchical root already set") unless value.nil?
+        required(:hierarchy_root).value(:nil?)
```
🟢 nice!

Kharonus commented on app/contracts/custom_fields/hierarchy/update_item_contract.rb:
```diff
@@ -0,0 +1,54 @@
+
+      rule(:item) do
+        key.failure("must exist") if value.new_record?
+        key.failure("must not be a root item") if value.root?
```
❓ is `root?` an extension method from `closure_tree`? Or can't I simply find it in our codebase?

mereghost replied:
[Yep](https://github.com/ClosureTree/closure_tree/tree/master?tab=readme-ov-file#instance-methods), from closure_tree.

Kharonus commented on app/services/custom_fields/hierarchy/hierarchical_item_service.rb:
```diff
@@ -55,25 +54,64 @@ def insert_item(parent:, label:, short: nil)
+      end
+
+      def move_item(item:, new_parent:)
+        Success(new_parent.append_child(item))
```
🟢 seems like closure tree indeed comes with some nice helpers.

mereghost replied:
Nice, huh?

Kharonus reviewed (approved):
LGTM generally, can you answer the last two questions though?

mereghost committed:
Remove instance variable, adds some docs, wraps some return values.
